### ISSUE TITLE
Implement function guards

### ISSUE THREAD

**Issue by lastmjs** [demergent-labs/azle#859]
Add support for guard functions to all of our custom "decorator" annotations as shown in the following example:

```ts
import { GuardResult } from "azle";

$query({guard: "guard_function"})
export function some_endpoint(): {
   return "string"
}

function guard_function(): GuardResult {}

```

All "decorators" (`$init`, `$pre_upgrade`, `$query`, etc.) will all accept an options object that has a "guard" property of type `String`. Canister authors can put the name of another global function in as a guard. The referenced function does not need to be exported nor a canister method.

If the guard function returns an `Err` result or throws, the annotated method should not continue executing. If the method returns something other than a result then the rust function should trap saying there was an unexpected type. The only way for the function to proceed is to return an `{ok: null}` value.

Tasks:

- [x] Parse the custom azle decorator annotations. Both idents and callables with/without guards.
- [x] Ensure all invalid cases have error messages that make sense and a correct origin.
- [x] Implement for all decorators (except init and post_upgrade)
- [x] Remove all muts
- [x] Parse the actual guard functions into a list
- [x] Move all related code into the new src/canister_methods directory
- [x] (Optional) replace all `AzleSomeTsType`s with `SourceMapped<SomeTsType>`
- [x] Tests for good and bad declarations
- [x] Tests for system methods
- [x] (Optional) improve the error messages
- [ ] (Optional) Re-write algorithm to not be look-behind, but rather look-ahead
- [x] Create an issue for the issue described below in the rust-cdk repo.
- [x] Update readme to explain that init and post_upgrade can't have guard functions right now. Also, remove the actual functionality so that it doesn't work. This should be in both the reference and canister life cycle sections of the book. Or create a ticket for Jordan to do that. See https://github.com/demergent-labs/azle/issues/954 and https://github.com/demergent-labs/azle/issues/939
- [x] ~~Either require all guard functions to be exported~~, or figure out how to call them at runtime. Function guards do not need to be exported, but there can't be two with the same name. Also, canister methods still need to be exported at the top level.
- [x] Update examples to camelCase

Resources:
* https://docs.rs/ic-cdk/latest/ic_cdk/attr.post_upgrade.html

**Comment by dansteren** [comment]
This is on hold pending decisions on https://github.com/demergent-labs/azle/issues/24

**Comment by dansteren** [comment]
We are unblocked on this issue. We've tweaked how we'll implement them a little bit but our new custom "decorator" annotations will allow us to implement this.

**Comment by lastmjs** [comment]
Hey, can you please add tests for the system methods into your guard functions example?

**Comment by dansteren** [comment]
Edge case to consider:

Consider the following two files:

_a.ts_
```a.ts
function commonlyNamedGuard(): GuardResult {
  return {ok: null}
}

$query({guard: commonlyNamedGuard});
export function methodA() {
  return true;
}
```

_b.ts_
```b.ts
function commonlyNamedGuard(): GuardResult {
  return {err: "This has the same name, but different functionality"}
}

$query({guard: commonlyNamedGuard});
export function methodB() {
  return true;
}
```
The expected behavior is that calling `methodA` would return `true` while calling `methodB` would return a level 4 error code. I.e., we need to check for method names within each file.

However, I think we would also like the following to work. If we have:

_a.ts_
```a.ts
import { guardFunction } from './b.ts';

$query({guard: guardFunction});
export function methodA() {
  return true;
}
```

_b.ts_
```b.ts
import { GuardResult } from "azle";

export function guardFunction(): GuardResult {
  return { ok: null }
}
```
We would also expect this to work.

These two behaviors seem at odds because we don't do anything dynamically...
